### DISCUSSION TITLE
Storing unique Terms in containers by equivalence (C++)

### DISCUSSION THREAD

**Discussion by jefftrull** [cvc5/cvc5#10264] [Q&A] [answered]
For my application I need to determine whether I have previously generated an equivalent Term (that is, the XOR of the two Terms is unsatisfiable). A linear search through all previously generated Terms for equivalence would work, but is painfully slow. A set-like structure (`std::unordered_set` or `std::set`) has much faster lookup, but Term doesn't have the right semantics to be used with one. It provides the hash `unordered_set` needs but there's no guarantee that two equivalent Terms will hash to the same value. A similar issue holds for `std::set` w.r.t. `operator<`.

Is there an efficient way to maintain a unique (by equivalence) collection of Terms?

**Comment by barrettcw** [comment, marked as answer]
This is a good question.  We solved a similar problem in Section 3 of this paper: https://theory.stanford.edu/~barrett/pubs/NRB+19.pdf

The short answer is to use a trie of sample points to index into your Term database and only do a full equivalence check if the two terms agree on all sample points.

**Reply by jefftrull** [reply]
Thank you for this excellent suggestion. I haven't fully incorporated it, or checked its performance, but I've built a prototype and it looks promising. If anyone is interested: https://gist.github.com/jefftrull/2ed4f54adcbe9468910c096c765742fe
